src/basic/capability-util.c

```diff
@@ -107,11 +107,13 @@ int capability_ambient_set_apply(uint64_t set, bool also_inherit) {
         unsigned long i;
         int r;
 
+        /* Add the capabilities to the ambient set (an possibly also the inheritable set) */
+
         /* Check that we can use PR_CAP_AMBIENT or quit early. */
         if (!ambient_capabilities_supported())
-                return 0;
-
-        /* Add the capabilities to the ambient set. */
+                return (set & all_capabilities()) == 0 ?
```

Comment:
Okay, so I see here we want to be more user friendly and let them know they are using an unsupported capability. That seems reasonable. One possible issue I see here is we can still mask errors, since `lookup_cap(..)` pulls in caps on newer kernels, and `all_capabilities()` uses the max capability on the current system. What do you think about just returning `-EOPNOTSUPP` if `set != 0`?

Reply:
I guess this also depends on what `cap_last_cap` is on these older systems. If it hasn't changed then my comment probably doesn't matter. I don't have a 3.x.x kernel to test either.

Reply:
So I am masking by all_capabilities() here on purpose, i.e. ignoring caps that do not exist locally. That's because if the cap doesn't exist at all, then we know for sure that lying about whether you got it or you don't doesn't have any effect, since what it protects doesn't exist either...  so in the interest of graceful compatibility i figured we shouldn't complain about caps that do not exist. i mean, think about it, if CAP_BLOCK_SUSPEND, CAP_WAKE_ALARM, or CAP_AUDIT_READ (as an example, i.e. the three most recently added caps) do not exist, then any attempts to make use of the concepts these caps protect is going to fail with EOPNOTSUPP or something similar anyway, so there's no need to refuse things...